### PR TITLE
Build Windows compatible font

### PR DESCRIPTION
**Pull request by RoelN**
This PR will have `make` output two fonts:

`NotoColorEmoji.ttf`, the unchanged regular font.

`NotoColorEmoji_WindowsCompatible.ttf`, which is the regular font, plus a `glyf` table with empty glyphs for each emoji, and a cmap format 4 3/1 table. This is needed by Windows 10 to recognise the font as valid font. This adds about ~8KB to the font.

This fixes https://github.com/googlefonts/noto-emoji/issues/43.

Note that the regular `NotoColorEmoji.ttf` is slightly bigger then the one [currently in the fonts directory](https://github.com/googlefonts/noto-emoji/blob/master/fonts/NotoColorEmoji.ttf). This appears to be because of how the PNG images are compressed. I'm using the tooling and instructions from the repo.

### CODE REVIEW

File: add_glyphs.py [hunk 1]
```diff
@@ -158,10 +161,17 @@ def add_glyph_data(font, seqs, seq_to_advance, vadvance):
...
+  # If a glyf table is present, empty glyphs will be added to ensure
+  # compatibility with systems requiring a glyf table, like Windows 10.
 
   cmap = get_font_cmap(font)
   hmtx = font['hmtx'].metrics
   vmtx = font['vmtx'].metrics
+  if 'glyf' in font:
+    pen = TTGlyphPen(None)
+    empty_glyph = pen.glyph()
+    glyf = font['glyf']
```
Comment (anthrotype) [inline add_glyphs.py:174]: I would add an `else: glyf = None` after this, then later when you add the empty glyphs you check `if glyf is not None:`.

File: add_glyphs.py [hunk 2]
```diff
@@ -327,14 +339,33 @@ def add_ligature(lookup, cmap, seq, name):
...
+  newtable.platformID = 3
+  newtable.platEncID = 1
+  newtable.language = 0
+  newtable.cmap = {}
 
-def update_font_data(font, seq_to_advance, vadvance, aliases):
+  # Format 4 only has unicode values 0x0000 to 0xFFFF
+  for cp, name in cmap.items():
+    if cp < 65535:
+      newtable.cmap[cp] = name
```
Comment (anthrotype) [inline add_glyphs.py:357]: you could make this in one line with a dict comprehension

```python
newtable.cmap = {cp: name for cp, name in cmap.items() if cp <= 0xFFFF)
```

File: add_glyphs.py [hunk 3]
```diff
@@ -327,14 +339,33 @@ def add_ligature(lookup, cmap, seq, name):
...
+  newtable = CmapSubtable.newSubtable(4)
+  newtable.platformID = 3
+  newtable.platEncID = 1
+  newtable.language = 0
+  newtable.cmap = {}
 
-def update_font_data(font, seq_to_advance, vadvance, aliases):
+  # Format 4 only has unicode values 0x0000 to 0xFFFF
+  for cp, name in cmap.items():
+    if cp < 65535:
```
Comment (anthrotype) [inline add_glyphs.py:356]: Should this be `<= 65535`?

**Review by anthrotype — COMMENTED** [review]
Thanks Roel. I left some comments below.

What's in "NotoColorEmoji_WindowsCompatible.tmpl.ttx.tmpl"? How does it differ from the other template ttx? Can you reuse the same template and add additional data via code?

File: add_glyphs.py (hunk 1 above)
Comment (RoelN) [reply add_glyphs.py:174]: Good call, thanks!

File: add_glyphs.py (hunk 3 above)
Comment (RoelN) [reply add_glyphs.py]: Well spotted!

**Comment by RoelN** [comment]
@anthrotype Thanks for the review! I addressed your comments. In regards to "Can you reuse the same template and add additional data via code?", I'm now re-using the original template and adding the `glyf` and `loca` during build, similar to how the new `cmap` is added. This is indeed cleaner, thanks for the pointer.

File: add_glyphs.py [hunk 4]
```diff
@@ -163,6 +167,16 @@ def add_glyph_data(font, seqs, seq_to_advance, vadvance):
...
 
+  # Add glyf table so empty glyphs will be added to ensure compatibility
+  # with systems requiring a glyf table, like Windows 10.
+  if add_glyf:
+    pen = TTGlyphPen(None)
+    empty_glyph = pen.glyph()
+    font['loca'] = _l_o_c_a.table__l_o_c_a()
+    font['glyf'] = glyf_table = _g_l_y_f.table__g_l_y_f()
+    glyf_table.glyphOrder = font.getGlyphOrder()
+    glyf_table.glyphs = {g:empty_glyph for g in glyf_table.glyphOrder}
```
Comment (anthrotype) [inline add_glyphs.py:178]:
```suggestion
    glyf_table.glyphs = {g: empty_glyph for g in glyf_table.glyphOrder}
```

File: add_glyphs.py [hunk 5]
```diff
@@ -163,6 +167,16 @@ def add_glyph_data(font, seqs, seq_to_advance, vadvance):
   hmtx = font['hmtx'].metrics
   vmtx = font['vmtx'].metrics
 
+  # Add glyf table so empty glyphs will be added to ensure compatibility
+  # with systems requiring a glyf table, like Windows 10.
+  if add_glyf:
+    pen = TTGlyphPen(None)
+    empty_glyph = pen.glyph()
+    font['loca'] = _l_o_c_a.table__l_o_c_a()
```
Comment (anthrotype) [inline add_glyphs.py:175]: you could also use the newTable function from fontTools.ttLib, which returns a new empty instance of table with given tag, so you don't need to import the table module with the long names, e.g. `newTable("loca")` or `newTable("glyf")`

File: add_glyphs.py [hunk 6]
```diff
@@ -327,14 +342,30 @@ def add_ligature(lookup, cmap, seq, name):
...
+
+def update_font_data(font, seq_to_advance, vadvance, aliases, add_cmap4, add_glyf):
   """Update the font's cmap, hmtx, GSUB, and GlyphOrder tables."""
   seqs = get_all_seqs(font, seq_to_advance)
-  add_glyph_data(font, seqs, seq_to_advance, vadvance)
+  add_glyph_data(font, seqs, seq_to_advance, vadvance, add_glyf)
   add_aliases_to_cmap(font, aliases)
   add_ligature_sequences(font, seqs, aliases)
-
+  if(add_cmap4):
```
Comment (anthrotype) [inline add_glyphs.py:367]:
```suggestion
  if add_cmap4:
```

**Review by anthrotype — APPROVED** [review]
LGTM, with minor comments

File: add_glyphs.py (hunk 5 above)
Comment (RoelN) [reply add_glyphs.py:175]: 👍 Thanks for the tip!

**Comment by RoelN** [comment]
@anthrotype Thanks again for the pointers. All your suggestions have been implemented!

**Comment by RoelN** [comment]
@anthrotype @rsheeter As okayed in the email thread, this can be merged. I don't have the rights to do that — could you do this? Thanks!

**Comment by anthrotype** [comment]
Thank you, apologies for the delay

**Comment by DeeDeeG** [comment]
A little late to suggest this, but I think the `glyf` table should probably only have one glyph for something obscure and unlikely to ever be used. Then it won't override real emoji characters in apps that don't support `CBLC/CBDT`.

But that implies some additional testing to make sure it works, and to confirm it's an actual improvement. I'm not 100% sure how to test, so I can't be totally confident this suggestion is right. Just wanted to mention the possibility. Maybe this sounds right to maintainers here?

Anyhow thanks for doing this PR!

**Comment by RoelN** [comment]
@DeeDeeG I tried that by having only the `.notdef` present, but then the font wouldn't compile with ttx. Either the `glyf` table has to be omitted completely, or it should contain every single glyph. But I agree that'd be cleanest, if that'd work!

**Comment by anthrotype** [comment]
note that having one glyph only or many identical glyphs doesn't make the glyph table shorter or longer. The loca will point to the same offset. The maxp table needs to contain the true number of glyphs, and be in agreement with all the tables in the font that contain glyphs (glyf or CBLC)

**Comment by DeeDeeG** [comment]
Thanks for the replies.

**Pull request merged**